samza-azure/src/main/java/org/apache/samza/system/azureblob/avro/AzureBlobOutputStream.java

```diff
@@ -230,6 +231,10 @@ public synchronized void releaseBuffer() throws IOException {
     }
   }
 
+  public synchronized void incrementNumberOfRecordsInBlob() {
```

Comment:
I suggest adding some javadocs here to clarify how this needs to be used (i.e. only for passing metadata for the blob when closing the blob).
Since `DataFileWriter` may still be buffering some data, this `totalNumberOfRecordsInBlob` may not reflect the exact number of records in the output stream until flush/close are actually called. Also, it would be good to clarify that this needs to be atomically incremented along with appending the record to the `DataFileWriter` (or else the final record count may not be correct when closing the blob). It looks like you are doing the atomic update already in `AzureBlobAvroWriter`, but the general management of writers and output streams is a bit complex, so it would be good to add some more details.